File: src/main/scala/com/kotori316/fluidtank/integration/ae2/TankAE2Plugin.java
package com.kotori316.fluidtank.integration.ae2;

import appeng.api.IAEAddonEntrypoint;
import appeng.api.networking.security.IActionSource;
import appeng.api.storage.IStorageMonitorableAccessor;
import appeng.api.storage.MEStorage;
import net.fabricmc.loader.api.FabricLoader;

import com.kotori316.fluidtank.FluidTank;
import com.kotori316.fluidtank.ModObjects;
import com.kotori316.fluidtank.tiles.TileTank;

public final class TankAE2Plugin implements IAEAddonEntrypoint {
    // public static final ResourceLocation LOCATION = new ResourceLocation(ModTank.modID, "attach_ae2");

    @Override
    public void onAe2Initialized() {
        if (FabricLoader.getInstance().isModLoaded("ae2") && FluidTank.config.enableAE2Integration) {
            CapHandler.event();
        }
    }

    static class CapHandler {
        public static void event() {
            IStorageMonitorableAccessor.SIDED.registerForBlockEntities((blockEntity, context) -> {
                if (blockEntity instanceof TileTank tank) return new AEConnectionCapabilityProvider(tank);
                else return null;
            }, ModObjects.TANK_TYPE(), ModObjects.TANK_VOID_TYPE(), ModObjects.TANK_CREATIVE_TYPE());
        }
    }

    static class AEConnectionCapabilityProvider implements IStorageMonitorableAccessor {
        private final TileTank tank;
        private AEFluidInv aeFluidInv;

        public AEConnectionCapabilityProvider(TileTank tank) {
            this.tank = tank;
        }

        @Override
        public MEStorage getInventory(IActionSource iActionSource) {
            if (aeFluidInv == null) aeFluidInv = new AEFluidInv(tank);
            return aeFluidInv;
        }
    }
}

File: src/main/scala/com/kotori316/fluidtank/integration/ae2/TankAE2Plugin.scala
package com.kotori316.fluidtank.integration.ae2

import com.kotori316.fluidtank.FluidTank
import com.kotori316.fluidtank.tiles.TileTank
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.event.AttachCapabilitiesEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.ModList

object TankAE2Plugin {
  val LOCATION = new ResourceLocation(FluidTank.modID, "attach_ae2")

  def onAPIAvailable(): Unit = {
    if (ModList.get.isLoaded("ae2"))
      MinecraftForge.EVENT_BUS.register(new CapHandler)
  }

  private class CapHandler {
    //noinspection ScalaUnusedSymbol
    @SubscribeEvent
    def event(event: AttachCapabilitiesEvent[BlockEntity]): Unit = {
      event.getObject match {
        case tank: TileTank =>
          val provider = new AEConnectionCapabilityProvider(tank)
          event.addCapability(LOCATION, provider)
        case _ =>
      }
    }
  }
}
